File: src/main/java/io/bloco/faker/components/Number.java
package io.bloco.faker.components;

import io.bloco.faker.FakerComponent;
import io.bloco.faker.FakerData;

public class Number extends FakerComponent {

    private static final int DEFAULT_DECIMAL_PART_DIGITS = 2;
    private static final int DEFAULT_FROM = 1;
    private static final int DEFAULT_TO = 5000;

    public Number(FakerData data) {
        super(data);
    }

    public String number(int digits) {
        String num = "";
        if (digits > 1) {
            num = nonZeroDigit();
            digits -= 1;
        }
        return num + leadingZeroNumber(digits);
    }

    public String leadingZeroNumber(int digits) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < digits; i++) {
            sb.append(digit());
        }
        return sb.toString();
    }

    public String decimalPart(int digits) {
        String num = "";
        if (digits > 1) {
            num = nonZeroDigit();
            digits -= 1;
        }
        return leadingZeroNumber(digits) + num;
    }

    public String decimal(int leftDigits) {
        return decimal(leftDigits, DEFAULT_DECIMAL_PART_DIGITS);
    }

    public String decimal(int leftDigits, int rightDigits) {
        return number(leftDigits) + "." + decimalPart(rightDigits);
    }

    public String digit() {
        return Integer.toString(randomHelper.number(10));
    }

    public String nonZeroDigit() {
        return Integer.toString(randomHelper.number(9) + 1);
    }

    public String hexadecimal(int digits) {
        if (digits < 1) {
            return "";
        }
        long num = randomHelper.number((long) Math.pow(16, digits));
        return String.format("%0" + digits + "x", num);
    }

    public int between() {
        return between(DEFAULT_FROM, DEFAULT_TO);
    }

    public int between(int from, int to) {
        return randomHelper.range(from, to);
    }

    public long between(long from, long to) {
      return randomHelper.range(from, to);
    }

    public double between(double from, double to) {
        return randomHelper.range(from, to);
    }

    public int positive() {
        return positive(DEFAULT_FROM, DEFAULT_TO);
    }

    public int positive(int from, int to) {
        return Math.abs(between(from, to));
    }

    public long positive(long from, long to) {
        return Math.abs(between(from, to));
    }

    public double positive(double from, double to) {
        return Math.abs(between(from, to));
    }

    public int negative() {
        return negative(DEFAULT_FROM, DEFAULT_TO);
    }

    public int negative(int from, int to) {
        return Math.abs(between(from, to)) * -1;
    }

    public long negative(long from, long to) {
        return Math.abs(between(from, to)) * -1;
    }

    public double negative(double from, double to) {
        return Math.abs(between(from, to)) * -1;
    }
}

File: src/main/kotlin/io/bloco/faker/components/Number.kt
package io.bloco.faker.components

import io.bloco.faker.FakerComponent
import io.bloco.faker.FakerData
import io.bloco.faker.helpers.RandomHelper
import kotlin.math.abs
import kotlin.math.pow

class Number(data: FakerData) : FakerComponent(data) {
    fun number(digits: Int): String {
        var tempDigits = digits
        var num = ""
        if (tempDigits > 1) {
            num = nonZeroDigit()
            tempDigits -= 1
        }
        return num + leadingZeroNumber(tempDigits)
    }

    fun leadingZeroNumber(digits: Int): String {
        return (1..digits).joinToString("") { digit() }
    }

    fun decimalPart(digits: Int): String {
        return leadingZeroNumber(digits - 1) + nonZeroDigit()
    }

    fun decimal(leftDigits: Int, rightDigits: Int = DEFAULT_DECIMAL_PART_DIGITS): String {
        return number(leftDigits) + "." + decimalPart(rightDigits)
    }

    fun digit(): String {
        return RandomHelper.number(10).toString()
    }

    fun nonZeroDigit(): String {
        return (RandomHelper.number(9) + 1).toString()
    }

    fun hexadecimal(digits: Int): String {
        if (digits < 1) {
            return ""
        }
        val num = RandomHelper.number(16.0.pow(digits.toDouble()).toLong())
        return "%0${digits}x".format(num)
    }

    @JvmOverloads
    fun between(from: Int = DEFAULT_FROM, to: Int = DEFAULT_TO): Int {
        return RandomHelper.range(from, to)
    }

    fun between(from: Long, to: Long): Long {
        return RandomHelper.range(from, to)
    }

    fun between(from: Double, to: Double): Double {
        return RandomHelper.range(from, to)
    }

    @JvmOverloads
    fun positive(from: Int = DEFAULT_FROM, to: Int = DEFAULT_TO): Int {
        return abs(between(from, to))
    }

    fun positive(from: Long, to: Long): Long {
        return abs(between(from, to))
    }

    fun positive(from: Double, to: Double): Double {
        return abs(between(from, to))
    }

    @JvmOverloads
    fun negative(from: Int = DEFAULT_FROM, to: Int = DEFAULT_TO): Int {
        return abs(between(from, to)) * -1
    }

    fun negative(from: Long, to: Long): Long {
        return abs(between(from, to)) * -1
    }

    fun negative(from: Double, to: Double): Double {
        return abs(between(from, to)) * -1
    }

    companion object {
        private const val DEFAULT_DECIMAL_PART_DIGITS = 2
        private const val DEFAULT_FROM = 1
        private const val DEFAULT_TO = 5000
    }
}
